app.ts

```diff
@@ -0,0 +1,140 @@
+}
+
+// converts strings into array and table
+function parseText(textString) {
+    let outputArray = textString.replace(/'/g, '""');
+    outputArray = JSON.parse(textString);
+    return outputArray;
+}
+
+class tableGenerator {
+    element: HTMLElement;
+    myTable: HTMLElement;
+
+    constructor(element: HTMLElement, firstRecord) {
+        this.element = element;
+        this.myTable = document.createElement('table');
+        this.element.appendChild(this.myTable);
+
+        // get table height based on  remaining window size after buttons
+
+        let contentHeight = (window.innerHeight - CONTROLS_HEIGHT);
+        this.myTable.setAttribute("height", "contentHeight");
+
+        // calculate row number based on table header + row heights + borders in CSS and ensure its never less than 1
+        let numRows = Math.floor((contentHeight - CONTROLS_HEIGHT - 26) / 46);
+        numRows = numRows <= 0 ? 1 : numRows;
+
+        // set number of records to get from server and ensure that the last record doesnt exceed the limit.
+        let endRecord = (firstRecord + numRows - 1);
+        endRecord = (endRecord >= RECORD_COUNT ? RECORD_COUNT - 1 : endRecord);
+
+        // get data from server
+        const columnsString = getData("http://localhost:2050/columns");
+        let recordsText = getData("http://localhost:2050/records?from=" + firstRecord + "&to=" + endRecord);
+
+        // parse text from api call into arrays
+        const columnsArray = parseText(columnsString);
+        let recordsTable = parseText(recordsText);
+
+        // write html for table header
+        for (const i of columnsArray) {
+            let th = document.createElement('th');
+            this.myTable.appendChild(th);
+
+            th.appendChild(document.createTextNode(" " + i));
+        }
+
+        // write html for table
+        for (const iEntry of recordsTable) {
+
+            let tr = document.createElement('tr');
+            this.myTable.appendChild(tr);
+
+            for (const jEntry of iEntry) {
+
+                let td = document.createElement('td');
+                td.appendChild(document.createTextNode(" " + jEntry));
+                tr.appendChild(td);
+            }
+        }
+    }
+}
+
+function createPage(firstRecord) {
+
+    // ensure that the user input is a number and within range of available records
+    if (firstRecord >= RECORD_COUNT || firstRecord < 0 || isNaN(firstRecord)) {
+        alert('please select a NUMBER between 0 and ' + (RECORD_COUNT - 1));
+        return;
+    }
+
+    let contentsDiv = this.content;
+    contentsDiv.innerHTML = ''; // empty table div to fill in a new table
+    let tableElement = new tableGenerator(contentsDiv, parseInt(firstRecord)); // create table using the tableGenerator class
+
+    // keep the ID of the first and last table entries to use for next and previous page buttons
+    LAST_ENTRY = contentsDiv.lastChild.lastChild.firstChild.textContent;
```

Comment:
This is very obscure. I'd rather get the div by id if you can so that the code is more readable